### PR TITLE
improvements to stack

### LINKED ISSUE #119: Accomodate `Base.stack`
https://github.com/JuliaLang/julia/pull/43334 recently landed, so now `using MLUtils` causes issues downstream in Flux. Not sure which level of the stack is better to address this in.

### LINKED ISSUE #121: stack can cause StackOverflowError or Segmentation fault on macOS ARM
Stacking vectors of length larger than 1353 causes a StackOverflowError:

```julia
julia> using MLUtils
for j in 1:2000
           a = [[1] for i in 1:j];
           stack(a, 2);
           @info j
end
```

```julia
.
.
.
[ Info: 1350
[ Info: 1351
[ Info: 1352
[ Info: 1353
ERROR: StackOverflowError:
Stacktrace:
 [1] __cat_offset!(::Matrix{Int64}, ::Tuple{Int64, Int64}, ::Tuple{Bool, Bool}, ::Tuple{Int64, Int64}, ::Matrix{Int64}, ::Matrix{Int64}, ::Vararg{Matrix{Int64}}) (repeats 1339 times)
   @ Base ./abstractarray.jl:1749
 [2] __cat(::Matrix{Int64}, ::Tuple{Int64, Int64}, ::Tuple{Bool, Bool}, ::Matrix{Int64}, ::Vararg{Matrix{Int64}})
   @ Base ./abstractarray.jl:1744
 [3] _cat_t(::Int64, ::Type{Int64}, ::Matrix{Int64}, ::Vararg{Matrix{Int64}})
   @ Base ./abstractarray.jl:1737
 [4] _cat(::Int64, ::Matrix{Int64}, ::Vararg{Matrix{Int64}})
   @ LinearAlgebra /Applications/Julia-1.8.app/Contents/Resources/julia/share/julia/stdlib/v1.8/LinearAlgebra/src/special.jl:424
 [5] cat(::Matrix{Int64}, ::Vararg{Matrix{Int64}}; dims::Int64)
   @ Base ./abstractarray.jl:1916
 [6] stack(xs::Vector{Vector{Int64}}; dims::Int64)
   @ MLUtils ~/.julia/packages/MLUtils/aThik/src/utils.jl:101
 [7] stack(x::Vector{Vector{Int64}}, dims::Int64)
   @ MLUtils ./deprecated.jl:72
```

However, if I restart the REPL I have this behaviour:

```julia
julia> using Pkg; Pkg.activate(".")
  Activating project at `~/Documents/doctorado/issues/MLUtils_stack`

julia> using MLUtils

julia> a = [[1] for i in 1:1354]; stack(a, 2);
ERROR: StackOverflowError:
Stacktrace:
 [1] __cat_offset!(::Matrix{Int64}, ::Tuple{Int64, Int64}, ::Tuple{Bool, Bool}, ::Tuple{Int64, Int64}, ::Matrix{Int64}, ::Matrix{Int64}, ::Vararg{Matrix{Int64}})
   @ Base ./abstractarray.jl:1746
 [2] __cat_offset!(::Matrix{Int64}, ::Tuple{Int64, Int64}, ::Tuple{Bool, Bool}, ::Tuple{Int64, Int64}, ::Matrix{Int64}, ::Matrix{Int64}, ::Vararg{Matrix{Int64}}) (repeats 1339 times)
   @ Base ./abstractarray.jl:1749
 [3] __cat(::Matrix{Int64}, ::Tuple{Int64, Int64}, ::Tuple{Bool, Bool}, ::Matrix{Int64}, ::Vararg{Matrix{Int64}})
   @ Base ./abstractarray.jl:1744
 [4] _cat_t(::Int64, ::Type{Int64}, ::Matrix{Int64}, ::Vararg{Matrix{Int64}})
   @ Base ./abstractarray.jl:1737
 [5] _cat(::Int64, ::Matrix{Int64}, ::Vararg{Matrix{Int64}})
   @ LinearAlgebra /Applications/Julia-1.8.app/Contents/Resources/julia/share/julia/stdlib/v1.8/LinearAlgebra/src/special.jl:424
 [6] cat(::Matrix{Int64}, ::Vararg{Matrix{Int64}}; dims::Int64)
   @ Base ./abstractarray.jl:1916
 [7] stack(xs::Vector{Vector{Int64}}; dims::Int64)
   @ MLUtils ~/.julia/packages/MLUtils/aThik/src/utils.jl:101
 [8] stack(x::Vector{Vector{Int64}}, dims::Int64)
   @ MLUtils ./deprecated.jl:72

julia> a = [[1] for i in 1:1353]; stack(a, 2);
ERROR: StackOverflowError:
Stacktrace:
 [1] __cat_offset!(::Matrix{Int64}, ::Tuple{Int64, Int64}, ::Tuple{Bool, Bool}, ::Tuple{Int64, Int64}, ::Matrix{Int64}, ::Matrix{Int64}, ::Vararg{Matrix{Int64}}) (repeats 1352 times)
   @ Base ./abstractarray.jl:1749
 [2] __cat(::Matrix{Int64}, ::Tuple{Int64, Int64}, ::Tuple{Bool, Bool}, ::Matrix{Int64}, ::Vararg{Matrix{Int64}})
   @ Base ./abstractarray.jl:1744
 [3] _cat_t(::Int64, ::Type{Int64}, ::Matrix{Int64}, ::Vararg{Matrix{Int64}})
   @ Base ./abstractarray.jl:1737
 [4] _cat(::Int64, ::Matrix{Int64}, ::Vararg{Matrix{Int64}})
   @ LinearAlgebra /Applications/Julia-1.8.app/Contents/Resources/julia/share/julia/stdlib/v1.8/LinearAlgebra/src/special.jl:424
 [5] cat(::Matrix{Int64}, ::Vararg{Matrix{Int64}}; dims::Int64)
   @ Base ./abstractarray.jl:1916
 [6] stack(xs::Vector{Vector{Int64}}; dims::Int64)
   @ MLUtils ~/.julia/packages/MLUtils/aThik/src/utils.jl:101
 [7] stack(x::Vector{Vector{Int64}}, dims::Int64)
   @ MLUtils ./deprecated.jl:72

julia> a = [[1] for i in 1:1352]; stack(a, 2);
ERROR: StackOverflowError:

julia> a = [[1] for i in 1:1351]; stack(a, 2);

signal (11): Segmentation fault: 11
in expression starting at REPL[8]:1
_ZN4llvm16InstCombinerImpl15visitSelectInstERNS_10SelectInstE at /Applications/Julia-1.8.app/Contents/Resources/julia/lib/julia/libLLVM.dylib (unknown line)
Allocations: 3717522 (Pool: 3708463; Big: 9059); GC: 5
```
And Julia exits.

```julia
(MLUtils_stack) pkg> st
Status `~/Documents/doctorado/issues/MLUtils_stack/Project.toml`
  [f1d291b0] MLUtils v0.2.10
```

```julia
julia> versioninfo()
Julia Version 1.8.0
Commit 5544a0fab76 (2022-08-17 13:38 UTC)
Platform Info:
  OS: macOS (arm64-apple-darwin21.3.0)
  CPU: 8 × Apple M1 Pro
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-13.0.1 (ORCJIT, apple-m1)
  Threads: 1 on 6 virtual cores
```

### PR DESCRIPTION
**Pull request by CarloLucibello**
~~Fix #121 and add a `rrule` for stack.~~

Edit:
Remove the `stack` implementation in favor of the one from Base.  The change is **technically breaking** since we don't support anymore `dims > N+1` (see https://github.com/JuliaML/MLUtils.jl/issues/119#issuecomment-1229127997)

Fix #121  fix #119

### CODE REVIEW

**Review by darsnack — APPROVED** [review]

**Comment by darsnack** [comment]
Errors due to #119

**Comment by CarloLucibello** [comment]
I think it is better to wait for https://github.com/JuliaML/MLUtils.jl/pull/124 to be merged and then remove the `stack` implementation from here

**Comment by darsnack** [comment]
Based on the discussion in #119, it doesn't seem like `Base.stack`/`Compat.stack` are 1-1 replacements for `MLUtils.stack`. I'm mostly concerned about `MLUtils.stack(x, 2)` style conflicting with `Compat.stack(f, xs)`.

> I think it is better to wait for https://github.com/JuliaML/MLUtils.jl/pull/124 to be merged

You mean so we release both changes together? Or is there a connection to Tables.jl that I am missing?

**Comment by CarloLucibello** [comment]
> Based on the discussion in https://github.com/JuliaML/MLUtils.jl/issues/119, it doesn't seem like Base.stack/Compat.stack are 1-1 replacements for MLUtils.stack. I'm mostly concerned about MLUtils.stack(x, 2) style conflicting with Compat.stack(f, xs).

`MLUtils.stack(x, 2)` was deprecated already with v0.2 in favour of the keyword arg version, so we should be fine in removing it for v0.3. 

>> I think it is better to wait for https://github.com/JuliaML/MLUtils.jl/pull/124 to be merged

> You mean so we release both changes together? Or is there a connection to Tables.jl that I am missing?

ops, I mentioned the wrong PR, I meant https://github.com/JuliaDiff/ChainRules.jl/pull/681

**Comment by mcabbott** [comment]
This method should also call stack (if it cannot be removed):

https://github.com/JuliaML/MLUtils.jl/blob/855f95b436e5b9ebe9704486eb3dfdf5dad011ee/src/utils.jl#L332-L342

File: src/MLUtils.jl [hunk 1]
```diff
@@ -22,7 +22,11 @@ import NNlib
 @traitdef IsTable{X}
 @traitimpl IsTable{X} <- Tables.istable(X)
                       
-
+if VERSION < v"1.9.0-DEV.1163"
+    import Compat: stack
+else
+    import Base: stack
+end
```
Comment (mcabbott) [inline src/MLUtils.jl:29]: Can simplify, as this will get Base version when available:
```suggestion
using Compat: stack
```

File: Project.toml [hunk 2]
```diff
@@ -1,10 +1,11 @@
 name = "MLUtils"
 uuid = "f1d291b0-491e-4a28-83b9-f70985020b54"
 authors = ["Carlo Lucibello <carlo.lucibello@gmail.com> and contributors"]
-version = "0.3.0"
+version = "0.4.0"
 
 [deps]
 ChainRulesCore = "d360d2e6-b24c-11e9-a2a3-2a2ae2dbcce4"
+Compat = "34da2185-b29b-5c13-b0c7-acf172513d20"
```
Comment (mcabbott) [inline Project.toml:8]: Should probably bound Compat = "4.2". (Can't bound ChainRules but that's ok.)

**Review by mcabbott — APPROVED** [review]

**Pull request merged**